proof/invariant-abstract/RealTime_AI.thy

```diff
@@ -1138,11 +1158,27 @@ lemma refill_pop_head_no_fail[wp]:
   unfolding refill_pop_head_def
   by wpsimp
 
-lemma non_overlapping_merge_refills_no_fail[wp]:
+lemma update_refill_hd_no_fail[wp]:
+  "no_fail (\<lambda>s. sc_refills_sc_at (\<lambda>refills. refills \<noteq> []) sc_ptr s)
+            (update_refill_hd sc_ptr f)"
+  unfolding update_refill_hd_def refill_pop_head_def
+  apply wpsimp
+  by (clarsimp simp: obj_at_def sc_at_pred_n_def)
+
+lemma set_refill_hd_no_fail[wp]:
   "no_fail (\<lambda>s. sc_refills_sc_at (\<lambda>refills. refills \<noteq> []) sc_ptr s)
-            (non_overlapping_merge_refills sc_ptr)"
-  unfolding non_overlapping_merge_refills_def refill_pop_head_def
-  by (wpsimp simp: update_refill_hd_def wp: update_sched_context_wp)
+            (set_refill_hd sc_ptr f)"
+  unfolding set_refill_hd_def refill_pop_head_def
+  by wpsimp
+
+lemma merge_nonoverlapping_head_refill_no_fail[wp]:
+  "no_fail (\<lambda>s. sc_refills_sc_at (\<lambda>refills. 1 < length refills) sc_ptr s)
+            (merge_nonoverlapping_head_refill sc_ptr)"
```

Comment:
indent looks off here, should align with `(` on prev line. same for `update_refill_hd_no_fail`